packages/apollo-engine-reporting/src/agent.ts

```diff
@@ -302,16 +313,33 @@ export class EngineReportingAgent<TContext = any> {
     if (protobufError) {
       throw new Error(`Error encoding trace: ${protobufError}`);
     }
-    const encodedTrace = Trace.encode(trace).finish();
 
-    const signature = await this.getTraceSignature({
-      queryHash,
-      documentAST,
-      queryString,
-      operationName,
-    });
+    let statsReportKey: string;
+    // It's important to check parse failure first, since parse failures always
+    // indicate validation failure
+    if (gqlParseFailure) {
+      statsReportKey = `## GraphQLParseFailure`;
+      if (this.options.sendOperationDocumentsOnValidationFailure && queryString) {
+        trace.operationBodyOnValidationFailure = queryString;
+        trace.operationNameOnValidationFailure = operationName;
+      }
+    } else if (gqlValidationFailure || !documentAST) {
+      statsReportKey = `## GraphQLValidationFailure`;
+      if (this.options.sendOperationDocumentsOnValidationFailure && queryString) {
+        trace.operationBodyOnValidationFailure = queryString;
+        trace.operationNameOnValidationFailure = operationName;
+      }
+    } else {
+      const signature = await this.getTraceSignature({
+        queryHash,
+        documentAST,
```

Comment:
I don't get how this works: this variable is nullable but the parameter it's being passed to isn't?

Reply:
The null case is handled above in an `else if` branch, so here we know that `documentAST` is defined.